joint_crab/systematics.py

```diff
@@ -1,333 +1,98 @@
-        "alpha": (1, 5),
-        "beta": (0.001, 1),
-        "reference": (0.3, 30),
-        "z_fermi": (-0.3, 0.3),
-        "z_magic": (-0.3, 0.3),
-        "z_hess": (-0.3, 0.3),
-        "z_veritas": (-0.3, 0.3),
-        "z_fact": (-0.3, 0.3),
-    }
-    m = Minuit(
-        all_instruments_energy_scale_wstat,
-        amplitude=3e-12,
-        reference=1,
-        alpha=2.4,
-        beta=0.2,
-        z_fermi=0.0,
-        z_magic=0.0,
-        z_hess=0.0,
-        z_veritas=0.0,
-        z_fact=0.0,
-        fix_reference=True,
-        error_amplitude=1e-13,
-        error_alpha=0.01,
-        error_beta=0.001,
-        error_reference=0.0,
-        error_z_fermi=0.001,
-        error_z_magic=0.001,
-        error_z_hess=0.001,
-        error_z_veritas=0.001,
-        error_z_fact=0.001,
-        limit_amplitude=limits_dict["amplitude"],
-        limit_alpha=limits_dict["alpha"],
-        limit_beta=limits_dict["beta"],
-        limit_reference=limits_dict["reference"],
-        limit_z_fermi=limits_dict["z_fermi"],
-        limit_z_magic=limits_dict["z_magic"],
-        limit_z_hess=limits_dict["z_hess"],
-        limit_z_veritas=limits_dict["z_veritas"],
-        limit_z_fact=limits_dict["z_fact"],
-        print_level=0,
-        errordef=1,
-    )
 
-    m.migrad()
-    m.hesse()
-    return m
+class SystematicsSpectrumFit(SpectrumFit):
+    delta = {"fermi": 0.05, "magic": 0.15, "hess": 0.15, "veritas": 0.15, "fact": 0.15}
 
+    def __init__(self):
+        self._model = SystematicsSpetrumModel()
+        self.fits = {name: self.make_fit(name) for name in config.all_datasets}
 
-def compute_fit_results(minuit):
-    # write the output in a yaml format similar to the one used for the gammapy - sherpa fit
-    results = {}  # dict to be written as a yaml
-    parameters = []  # list of parameters
-    for name, value, error, fixed in zip(
-        minuit.values.keys(),
-        minuit.values.values(),
-        minuit.errors.values(),
-        minuit.fixed.values(),
-    ):
-        # the output with the modified likelihood does not have astropy units
-        unit = ""
-        if name == "amplitude":
-            unit = "1 / (cm2 s TeV)"
-        if name == "reference":
-            unit = "TeV"
-        # dictionary to be appended to parameters
-        param_dict = {
-            "name": name,
-            "value": value,
-            "unit": unit,
-            "frozen": fixed,
-            # "min": limits_dict[name][0],
-            # "max": limits_dict[name][1],
-            "error": error,
-        }
-        parameters.append(param_dict)
+    def make_fit(self, name):
+        dataset = config.datasets[name]
+        observations = dataset.get_SpectrumObservationList()
+        model = self._model.make_model()
+        fit_range = dataset.energy_range
+        return SpectrumFit(obs_list=observations, model=model, fit_range=fit_range)
 
-    results["parameters"] = parameters
+    def total_stat(self, parameters):
```

Comment:
Does `SpectrumFit` knows this is the quantity to minimize?

Reply:
Yes, that's how it works.